### PR TITLE
emit thumb: Add support for MRS instruction

### PR DESCRIPTION
**Pull request by HenrikSolver**
This change makes it possible to use the MRS instruction in asm_thumb code.

### CODE REVIEW

**Comment by dpgeorge** [comment]
I think "MRS" should be lower case to match all the other opcodes.  And perhaps the special reg names as well...

What is the reason you need access to these registers?

**Comment by PappaPeppar** [comment]
Regarding case: 
I were thinking that way as well, but in the end I just followed the convention in the documentation from ARM.

The reason is some experimentation described here http://forum.micropython.org/viewtopic.php?f=3&t=1164 . Then there have been some discussions about using the PRIMASK register (in #1555), which require the MSR instruction. Since I have done part of that job already I wanted to share it.

**Comment by dpgeorge** [comment]
I agree that loading special registers is useful in the context of inline assembler (and already we have vmrs to load floating point special regs).  But this patch here costs 300 bytes and likely only a few special regs are really needed.  So I suggest just supporting IPSR and PRIMASK to begin with, and implementing it the same way as vmrs.

mrs should be lower case, but the special reg names can stay upper case (that's how vmrs does it).

**Comment by dpgeorge** [comment]
Actually, I think we want BASEPRI, not PRIMASK.  PRIMASK can be changed using cpsie/cpsid.  From the docs about BASEPRI: "[if non zero] the processor does not process any exception with a priority value greater than or equal to BASEPRI".

And maybe we should make the special regs lower case, since normal regs are also lower case, and then it's least surprise for the user (just tell them everything is lower case).

**Comment by PappaPeppar** [comment]
I do not fully agree with you about mrs in lower case, but I will not argue about it if you want it this way :-). I have removed support for all registers exept IPSR and BASEPRI, updated the tests and rebased to latest master.

File: py/emitinlinethumb.c [hunk 1]
```diff
@@ -196,6 +203,20 @@ STATIC mp_uint_t get_arg_reg(emit_inline_asm_t *emit, const char *op, mp_parse_n
     return 0;
 }
 
+STATIC mp_uint_t get_arg_special_reg(emit_inline_asm_t *emit, const char *op, mp_parse_node_t pn) {
+    const char *reg_str = get_arg_str(pn);
+    for (mp_uint_t i = 0; i < MP_ARRAY_SIZE(special_reg_name_table); i++) {
+        const special_reg_name_t *r = &special_reg_name_table[i];
+        if (strncmp(r->name, reg_str, MAX_SPECIAL_REG_NAME_LENGTH) == 0) {
```
Comment (dpgeorge) [inline py/emitinlinethumb.c]: This will catch extra stuff, eg, reg_str = "BASEPRIXXX" will be matched.

File: tests/inlineasm/asmspecialregs.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import pyb
+import time
```
Comment (dpgeorge) [inline tests/inlineasm/asmspecialregs.py]: We should keep inlineasm tests free of non-core imports, like pyb and time.  So I'd say remove the test requiring pyb.Timer.  It's likely to be flaky anyway (and it's the reason why Travis fails, due to qemu-arm build not having pyb module).

**Comment by PappaPeppar** [comment]
I have now fixed the issues.

**Comment by dpgeorge** [comment]
Merged in e242b1785f9e25b36f5d225652c679cd7cec6ec0, thank you!

**Pull request closed without merging**